### PR TITLE
#4713 - Replace appcache with a service worker cache

### PR DESCRIPTION
**Pull request by kennsippell**
# Description 
This change replaces the underlying technology which allows the webapp to load while offline. This change removes the use of the legacy appcache and adds offline support via a PWA service worker. 
#4713

# Method
## Chrome 53 Constraint 
The technical approach taken in this proposed change is guided by a non-obvious constraint of service works in Chrome 53. Prior to Chrome 66, the web requests from the service worker don't include credentials (cookies, basic auth, etc). This means all resource we wish to cache must be accessible via unauthenticated endpoints. 

## Overview of my Approach

* **Service Worker Code** - I use the [swPrecache library from Google](https://github.com/GoogleChromeLabs/sw-precache#service-worker-precache) to generate the service worker script at build time. SwPrecache has been replaced by [Workbox](https://developers.google.com/web/tools/workbox/) which is Google's latest solution for generating service workers. Details as to why I'm using the legacy technology discussed in "swPrecache vs Workbox" below.  
* **New public endpoints** - When API starts, I extract all cacheable resources from _design/medic and write them to a new directory in `api/src/extracted`. So instead of serving resources through the authenticated endpoint `/medic/_design/medic/_rewrite/js/inbox.js`, the same resource is served from the public `/js/inbox.js`. 
* **Initial service worker installation** - I updated bootstrapper to handle the synchronous install of the service worker and show an error + reload action if precaching of any resource fails. (#3568) 
* **Triggering Updates** - A service worker update should be triggered when any resource in the cache changes (if not, you have to refresh twice to get a change). Since the `service-worker.js` generated at build time contains a hash of every cached resource, watching the script itself for changes is sufficient to detect when a cache update is necessary.  I re-use "ddoc extract" pattern to trigger the updates. 
* **Updating** - When the service worker updates, only the files which have changed are downloaded. If the service worker fails to download a required file for any reason (eg. connectivity), I retry on a timer every five minutes until it is successful. Only after the service worker updates is the user prompted to refresh. 
* **Scope** - A service worker can't have a "scope" broader than its own location. So although I serve the javascript from `/js/service-worker.js`, I create an alias at `/service-worker.js` to give it broad control to handle all caching concerns. 
* **Upgrading from AppCache to Service Workers** - If you login to current master (where resources are cached via appcache) and upgrade to this change - it is important to clear the old cache before caching anew. It is not possible to clear the appcache programmatically. I set `<html manifest="empty">` which requests an empty manifest and clears the cache. 
* **Rewire** - I've added a dependency on the rewire library for mocha testing. It's awesome and we should use it! 
 
## swPrecache vs Workbox 
It is worth checking out the [Support Details for swPrecache](https://github.com/GoogleChromeLabs/sw-precache#support). TLDR: "We would recommend Workbox for new projects" but "we’ll continue to support sw-precache". 
 
Workbox looks awesome. Most notably check out [enabling offline Google Analytics](https://developers.google.com/web/tools/workbox/modules/workbox-google-analytics). So why isn't this PR using Workbox? The short answer is we probably should – but this is a big PR and Workbox is going to complicate things.  
 
Here is why I'm using the swPrecache library: 
 
1. **The community support isn't perfect for either option** - Workbox 3 supports Chrome >51, but modern [Workbox 4 only supports Chrome >56](https://github.com/GoogleChrome/workbox/releases) so really both swPrecache and latests Workbox are [both not supporting us](https://www.youtube.com/watch?v=bVwh8PW6r88). 
1. **Partial Precache State** - SwPrecache is pretty much exactly what we want by default: it fails to install the service worker if any resource doesn't get precached and thus the possibility of a partial precached state. Workbox delivers a best effort precache and it isn't immediately obvious to me how to best avoid a partial precache state. 
1. **Dependent Scripts** - SwPrecache inlines all dependencies directly in the service worker so we just serve one script and we are done. Workbox has a modular build which deploys dependencies along-side your service worker (~20 files). A bit messy and not a perfect fit for our build script which ends up adding twenty more attachments to the database. A quick attempt to build only the required resources ended up with some dependencies being pulled from a cdn – which is problematic for projects using domain-based data contracts.  
 
All solvable problems, but a bit of untangling to do. Imo -- the best case would be using Workbox to output a script which does exactly what swPrecache is outputting now. I'm confident we can get there if we are interested in other features of Workbox *cough* - offline google analytics - *cough*. 
 
# Questions for Reviewer 

1. Let's discuss the general approach in this PR and maybe somebody can test a few scenarios on a device or two? If we like the result, maybe it is easier to carve this into chunks and review and submit in pieces. 
1. I just copied the list of resources to cache from appcache.manifest. We cache 29 files (some things which we maybe don't use). Maybe worth following up with an item to reduce how much we cache? 
1. The Application tab in devtools reports the disk use of appcache as 20.6MB and the same resources cached using service workers as 39.7MB. Unsure of the reason, but possible discussion point. 
1. It would be nice to get some guidance on exactly what horti scenarios would be good to test for this. 
1. Should we use the service worker to cache the favicon? 
1. Should we use the service worker to cache the login page? Even if you can't login, it's a better experience than seeing the non-localized white unbranded error page for login when offline. 
1. Upgrading from master to this change doesn't give the user the "Update" prompt because there is no change to a document being watched in master. Since this change won't be deployed atomically and will be instead packaged with other changes in v3.5 - I think that is safe. But we can consider tombstoning the legacy `appcache` document to make this change prompt an upgrade and also keep the database a bit cleaner. Looking for advice as I'm unsure how to best handle that. 
1. For performance reasons, we could consider making the service worker download and precache resources on the login page instead of after login. It is a one-line change which could benefit some users but possibly annoy others. 
1. Instead of using "ddoc-extraction," I considered calculating the hash of the service worker file at build-time and setting a single attribute on the medic-client ddoc. Think this would be cleaner? 
1. I put the core grunt logic into its own file instead of dumping it into the Gruntfile.js. I find that Gruntfile really hard to parse, and easily get lost in its large indentations. What do you think of starting to chop up that file into separate more manageable pieces (as done here)? 
1. Want to add a new UI state into the bootstrapper which says something like "Downloading App..." For the duration of the period in which we precache the service worker's resources? 
1. This change will block testing on insecure origins except localhost. This is pretty awkward for testing dev builds on a phone – I ran a DNS server + nginx proxy + vpn on my phone to work around it. This is likely something we should address in this change or shortly after. Looking to discuss the best approach. 
1. Is it safe the remove the write access controls to the deprecated appcache document in this change? 

# Scenarios I Tested

1. Incognito, login. Go offline. Refresh: user should get offline experience. 
1. Incognito, delete a required attachment, login: user should see an error dialog prompting to check connection. Rebuild and click retry: login should succeed and page should work offline. 
1. Incognito, login, update a css file, build, sync: user should see prompt to update. Update: user should see change in css appear. Network shows only the css file being updated. 
1. Incognito, login, update a css file, build, delete the css attachment in db, sync: user should see error in console and should not be prompted to update. Rebuild: after five minutes user should see prompt to update. Update: user should see change in css appear. 
1. If service-worker.js results in 404, then the app should not load and user should see error prompting to check connection. 
1. Incognito, login, update a css file, build, refresh without syncing: user should see prompt to update after refresh. Update: user should see change in css appear. 
1. Incognito, login to master (AppCache version). Rebuild to proposed change, sync. User should see prompt to update. 

# License
The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by kennsippell** [comment]
Tests are failing with this error:

```
Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
```

Accord to travis this is because "If your build relies on encrypted variables to run... You won’t be able to run these tests for pull requests from external contributors."
https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions

**Comment by garethbowen** [comment]
@kennsippell I've raised [this issue](https://github.com/medic/medic-webapp/issues/5198) to find a fix for the generic build issue.

Do you think this change is ready for review? Have you run the tests locally?

**Comment by kennsippell** [comment]
@gareth ya everything is passing locally. Would love a review!

**Comment by kennsippell** [comment]
Reopening as #5219
